### PR TITLE
session: add error log to catch bug stack in Commit

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Make the code more robust and easier to debug bug.

### What is changed and how it works?

Add an error log if the transaction state is unexpected when calling `Commit` 

### Check List 

Tests 

 - No code

Related changes

 - Need to cherry-pick to the release branch

@winkyao @crazycs520

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8687)

### CODE REVIEW

File: session/txn.go [hunk 1]
```diff
@@ -157,6 +158,15 @@ func (st *TxnState) Commit(ctx context.Context) error {
...
 		return errors.Trace(err)
 	}
+	if len(st.mutations) != 0 || len(st.dirtyTableOP) != 0 || st.buf.Len() != 0 {
+		log.Errorf("The code should never run here, TxnState=%#v, mutations=%#v, dirtyTableOP=%#v, buf=%#v something must be wrong: %s",
+			st,
+			st.mutations,
+			st.dirtyTableOP,
+			st.buf,
+			debug.Stack())
+		st.cleanup()
```
Comment (winkyao) [inline session/txn.go]: Should we return an error here？

Comment (tiancaiamao) [reply session/txn.go]: done

File: session/txn.go [hunk 2]
```diff
@@ -293,7 +310,11 @@ func (s *session) getTxnFuture(ctx context.Context) *txnFuture {
 
 	oracleStore := s.store.GetOracle()
 	tsFuture := oracleStore.GetTimestampAsync(ctx)
-	return &txnFuture{tsFuture, s.store}
+	ret := &txnFuture{future: tsFuture, store: s.store}
+	if x := ctx.Value("mockGetTSFail"); x != nil {
```
Comment (winkyao) [inline session/txn.go]: Can we use gofail to mock the error? To avoid hurting the performance.

Comment (tiancaiamao) [reply session/txn.go]: No we can't. `gofail` is works globally, 
 `create table t ...` will run some background job and hang in unit test

```
create table t ...
// gofail enable here, the test code hang
select * from t ...     
insert into t values ...
```

Comment (winkyao) [reply session/txn.go]: Could we use gofail + global variable?

Comment (winkyao) [reply session/txn.go]: I advice you using below code:

```
if globalVariableEnable && x := ctx.Value("mockGetTSFail"); x != nil {
```

Comment (tiancaiamao) [reply session/txn.go]: The performance is good in general:
https://golang.org/src/context/context.go#L183 @winkyao

Comment (winkyao) [reply session/txn.go]: Never mind.

**Pull request merged**